spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java

```diff
@@ -0,0 +1,309 @@
+
+      fileScanTasks.addAll(latestMicroBatch.tasks());
+    } while (latestMicroBatch.snapshotId() != endOffset.snapshotId());
+
+    return fileScanTasks;
+  }
+
+  private StreamingOffset getNextAvailableSnapshot(long snapshotId) {
+    Snapshot previousSnapshot = table.snapshot(snapshotId);
+    Snapshot pointer = table.currentSnapshot();
+    while (pointer != null && previousSnapshot.snapshotId() != pointer.parentId()) {
+      Preconditions.checkState(pointer.operation().equals(DataOperations.APPEND),
+          "Encountered Snapshot DataOperation other than APPEND.");
+
+      pointer = table.snapshot(pointer.parentId());
+    }
+
+    Preconditions.checkState(pointer != null,
+        "snapshot on which the stream operated has been garbage collected.");
+
+    return new StreamingOffset(pointer.snapshotId(), 0L, false);
+  }
+
+  interface OffsetLog {
+    static OffsetLog getInstance(SparkSession spark, String checkpointLocation) {
+      return new OffsetLogImpl(spark, checkpointLocation);
+    }
+
+    void addInitialOffset(StreamingOffset offset);
+
+    boolean isOffsetLogInitialized();
+
+    StreamingOffset getLatest();
+  }
+
+  private static class OffsetLogImpl implements OffsetLog {
+    private final IcebergSourceOffsetLog offsetSeqLog;
+
+    OffsetLogImpl(SparkSession spark, String checkpointLocation) {
+      this.offsetSeqLog = checkpointLocation != null ?
+          new IcebergSourceOffsetLog(spark, checkpointLocation) :
+          null;
+    }
+
+    @Override
+    public void addInitialOffset(StreamingOffset offset) {
+      this.offsetSeqLog.add(0, offset);
+    }
+
+    @Override
+    public boolean isOffsetLogInitialized() {
+      return offsetSeqLog != null &&
+          offsetSeqLog.getLatest() != null &&
+          offsetSeqLog.getLatest().isDefined();
+    }
+
+    @Override
+    public StreamingOffset getLatest() {
+      return offsetSeqLog.getLatest().get()._2;
+    }
+  }
+
+  private static class IcebergSourceOffsetLog extends HDFSMetadataLog<StreamingOffset> {
+    private final String path;
+
+    IcebergSourceOffsetLog(SparkSession sparkSession, String path) {
+      super(sparkSession, path, ClassTag.apply(StreamingOffset.class));
+      this.path = path;
+    }
+
+    @Override
+    public void serialize(StreamingOffset metadata, OutputStream out) {
+      try {
+        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(out, StandardCharsets.UTF_8));
+        writer.write(metadata.json());
+        writer.flush();
+      } catch (IOException ioException) {
+        throw new IllegalArgumentException(
+            String.format("Failed to deserialize latest checkpoint from: %s, with error: %s.", path, ioException));
+      }
+    }
+
+    @Override
+    public StreamingOffset deserialize(InputStream in) {
+      try {
+        String content = CharStreams.toString(new InputStreamReader(in, StandardCharsets.UTF_8));
```

Comment:
We're trying not to expand our use of Guava because it needs to be shaded and can change between versions. I think there should be an easy non-Guava way to do this. Can you try that instead?

Reply:
yeah - guava shading & cross version changes are painful. 
this checkstyle error lead me to Guava. followed your suggestion & removed this Guava dependency. thank you.
![image](https://user-images.githubusercontent.com/11813428/121097052-39051680-c7a8-11eb-815c-b98b1c12369b.png)